README.md
### Squareroot.js

A 2d/3d engine for WebGL and Canvas rendering and animation created by Bartek Drozdz. You can read all about it [here](http://sqr.holotype.co/), including API docs, tutorials and more!

License: 
All the code is licensed under [MIT](./LICENSE)


CHANGELOG.md

##### v3.0 b 99
*Mar 27th 2015*

Added this very Changelog file


##### v3.0 b 100
*Mar 31st 2015*

Added forward vector to Transform


##### v3.0 b 102
*Mar 31st 2015*

Shader.setUniform returns shader instance even if the uniform was not found - to avoid obsure errors


##### v3.0 b 103
*Apr 2nd 2015*

Added height inversion to V3.toScreenSpace


##### v3.0 b 104
*Apr 2nd 2015*

Adding option to slightly oversize skyboxes - maybe will help with seams on video files?


##### v3.0 b 105
*Apr 2nd 2015*

The previous update was a bad idea...


##### v3.0 b 106
*Apr 17th 2015*

Adding Quaternion.dot (static), quaternion.neg and a trick in slerp to make it always choose the shortest distance


##### v3.0 b 108
*Apr 17th 2015*

Unroll previous trick in Quaternion.slerp


##### v3.0 b 109
*Apr 17th 2015*

Texture does not generate mipmaps by default now, option mipmap=true needs to be specified


##### v3.0 b 110
*Apr 17th 2015*

Fix on Texture filtering related to mimpap setting above


##### v3.0 b 111
*Apr 18th 2015*

Adding that Quaternion.slerp trick, with a fix. This time hopefully it works!


##### v3.0 b 112
*May 6th 2015*

Creating a separate build for two+math


##### v3.0 b 113
*May 6th 2015*

CanvasRenderer can now be called without root (will just clear the canvas)


##### v3.0 b 114
*May 12th 2015*

CanvasRenderer will now clear correctly even with alpha in background


##### v3.0 b 115
*May 15th 2015*

Adding reversing normals option to icosphere


##### v3.0 b 116
*May 29th 2015*

name parameter was missing in Transform2d


##### v3.0 b 117
*Aug 8th 2015*

Adding SceneParser to primitives


##### v3.0 b 118
*Aug 8th 2015*

Adding perQuadUV option to plane contructor in primitives


##### v3.0 b 119
*Aug 12th 2015*

Adding small settings to Renderer and Plane + new demo


##### v3.0 b 123
*Sep 18th 2015*

Removing extras bucket, adding vr


##### v3.0 b 124
*Oct 6th 2015*

Adding Math.map


##### v3.0 b 125
*Oct 7th 2015*

Cleaning the Texture class a bit, update to API (all backwards compatible btw)


##### v3.0 b 128
*Oct 15th 2015*

Removing GLSL compilation from build process


##### v3.0 b 131
*Oct 16th 2015*

Added toCSS and lerp to SQR.Color


##### v3.0 b 132
*Oct 18th 2015*

Changing how disabling depth works in Transform. Not the entire DEPTH_TEST is disabled, not just the depth mask


##### v3.0 b 133
*Oct 27th 2015*

Very important: the welcome console message is now in color :)


##### v3.0 b 136
*Oct 29th 2015*

Started reworking the mesh creation functions (aka Primitives)


##### v3.0 b 137
*Nov 2nd 2015*

Changes in Renderer/Context classes and new Mesh API


##### v3.0 b 149
*Nov 9th 2015*

Adding builtin shaders


##### v3.0 b 150
*Nov 10th 2015*

Realizing that we need to rewrite the SceneParser from scratch. Otherwise, some minor fixes


##### v3.0 b 153
*Jan 29th 2016*

Recovering updates [1]


##### v3.0 b 154
*Mar 11th 2016*

Adding custom mouse position param to Pointer3d.fromMousePosition


##### v3.0 b 155
*Mar 29th 2016*

Shader is not mandatory when creating a postEffect


##### v3.0 b 172
*Apr 20th 2016*

Fix for multi-context cases

##### v3.0 b 175
*Jun 2nd 2016*

Removing gloabl fullscreenquad in PostEffect


##### v3.0 b 176
*Jun 22nd 2016*

Adding Matrix2D to Kuula build


##### v3.0 b 177
*Jun 22nd 2016*

Adding Framebuffer to Kuula build


##### v3.0 b 178
*Jun 22nd 2016*

Adding texture config options to FrameBuffer


##### v3.0 b 179
*Jul 18th 2016*

AddingCube to Kuula build


##### v3.0 b 180
*Aug 17th 2016*

Not adding listener to Gyro on start (lazy add instead) + added 'two' package to Kuula build


##### v3.0 b 181
*Aug 17th 2016*

Adding resolution parameter to CanvasRenderer.setSize


##### v3.0 b 182
*Aug 18th 2016*

/


##### v3.0 b 183
*Aug 18th 2016*

/


##### v3.0 b 184
*Aug 18th 2016*

/


##### v3.0 b 185
*Aug 18th 2016*

/


##### v3.0 b 186
*Aug 18th 2016*

/


##### v3.0 b 187
*Aug 18th 2016*

/


##### v3.0 b 188
*Aug 18th 2016*

/


##### v3.0 b 189
*Aug 18th 2016*

/


##### v3.0 b 190
*Aug 30th 2016*

Fixing texture creating in FrameBuffer


##### v3.0 b 191
*Aug 30th 2016*

Fixing texture creating in FrameBuffer [2]


##### v3.0 b 194
*Oct 29th 2016*

Adding texture premul alpha setting



##### v3.0 b 195
*Oct 29th 2016*

Adding context destroy function


##### v3.0 b 196
*Oct 29th 2016*

Trying to figure out the texture errors in FF


##### v3.0 b 197
*Nov 2nd 2016*

Testing different texture formsts for an iOS bug


##### v3.0 b 198
*Nov 2nd 2016*

Adding one option to FrameBuffer to leave all binded after creation


##### v3.0 b 199
*Nov 2nd 2016*

Testing different texture formsts for an iOS bug [2]


##### v3.0 b 200
*Mar 7th 2017*

Fixing Gyro to work ok on Desktops with no actual gyroscope


##### v3.0 b 208
*Apr 28th 2017*

Fixing a bug in Gyro on hybrind screens


##### v3.0 b 209
*Apr 28th 2017*

Fixing a bug in Gyro on hybrind screens


##### v3.0 b 211
*Oct 18th 2017*

Adding premultiply alpha option to textures


##### v3.0 b 212
*Oct 18th 2017*

Fixing bug where options passed to constructor would be ignored in textures with no source


##### v3.0 b 213
*Oct 18th 2017*

Adding Face to Kuula build


##### v3.0 b 214
*Oct 18th 2017*

Adding PLane insteas of Face to Kuula build


##### v3.0 b 215
*Oct 19th 2017*

Matrix after lookAt was not correctly preserving scale


##### v3.0 b 216
*Oct 25th 2017*

Enabling Gyro to accept external data


##### v3.0 b 217
*Oct 25th 2017*

Enabling Gyro to accept external data


##### v3.0 b 218
*Oct 30th 2017*

Adding collider and intersection to Kuula build


##### v3.0 b 219
*Dec 1st 2017*

Color CSS value now returns alpha


##### v3.0 b 220
*Dec 1st 2017*

Removed an annoying warning in Texture re; mipmaps


##### v3.0 b 221
*Dec 1st 2017*

Turns out that Color stuff was a bit more tricky


##### v3.0 b 222
*Feb 26th 2018*

Small fix to SceneParser


##### v3.0 b 223
*Feb 26th 2018*

Including scene folder in the main build


##### v3.0 b 224
*Mar 14th 2018*


lib/FastBlur.js
/*

Superfast Blur - a fast Box Blur For Canvas

Version: 	0.5
Author:		Mario Klingemann
Contact: 	mario@quasimondo.com
Website:	http://www.quasimondo.com/BoxBlurForCanvas
Twitter:	@quasimondo

In case you find this class useful - especially in commercial projects -
I am not totally unhappy for a small donation to my PayPal account
mario@quasimondo.de

Or support me on flattr:
https://flattr.com/thing/140066/Superfast-Blur-a-pretty-fast-Box-Blur-Effect-for-CanvasJavascript

Copyright (c) 2011 Mario Klingemann

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.
*/
var mul_table = [ 1,57,41,21,203,34,97,73,227,91,149,62,105,45,39,137,241,107,3,173,39,71,65,238,219,101,187,87,81,151,141,133,249,117,221,209,197,187,177,169,5,153,73,139,133,127,243,233,223,107,103,99,191,23,177,171,165,159,77,149,9,139,135,131,253,245,119,231,224,109,211,103,25,195,189,23,45,175,171,83,81,79,155,151,147,9,141,137,67,131,129,251,123,30,235,115,113,221,217,53,13,51,50,49,193,189,185,91,179,175,43,169,83,163,5,79,155,19,75,147,145,143,35,69,17,67,33,65,255,251,247,243,239,59,29,229,113,111,219,27,213,105,207,51,201,199,49,193,191,47,93,183,181,179,11,87,43,85,167,165,163,161,159,157,155,77,19,75,37,73,145,143,141,35,138,137,135,67,33,131,129,255,63,250,247,61,121,239,237,117,29,229,227,225,111,55,109,216,213,211,209,207,205,203,201,199,197,195,193,48,190,47,93,185,183,181,179,178,176,175,173,171,85,21,167,165,41,163,161,5,79,157,78,154,153,19,75,149,74,147,73,144,143,71,141,140,139,137,17,135,134,133,66,131,65,129,1];


var shg_table = [0,9,10,10,14,12,14,14,16,15,16,15,16,15,15,17,18,17,12,18,16,17,17,19,19,18,19,18,18,19,19,19,20,19,20,20,20,20,20,20,15,20,19,20,20,20,21,21,21,20,20,20,21,18,21,21,21,21,20,21,17,21,21,21,22,22,21,22,22,21,22,21,19,22,22,19,20,22,22,21,21,21,22,22,22,18,22,22,21,22,22,23,22,20,23,22,22,23,23,21,19,21,21,21,23,23,23,22,23,23,21,23,22,23,18,22,23,20,22,23,23,23,21,22,20,22,21,22,24,24,24,24,24,22,21,24,23,23,24,21,24,23,24,22,24,24,22,24,24,22,23,24,24,24,20,23,22,23,24,24,24,24,24,24,24,23,21,23,22,23,24,24,24,22,24,24,24,23,22,24,24,25,23,25,25,23,24,25,25,24,22,25,25,25,24,23,24,25,25,25,25,25,25,25,25,25,25,25,25,23,25,23,24,25,25,25,25,25,25,25,25,25,24,22,25,25,23,25,25,20,24,25,24,25,25,22,24,25,24,25,24,25,25,24,25,25,25,25,22,25,25,25,24,25,24,25,18];


function boxBlurImage( imageID, canvasID, radius, blurAlphaChannel, iterations ){

 	var img = document.getElementById( imageID );
	var w = img.naturalWidth;
    var h = img.naturalHeight;

	var canvas = document.getElementById( canvasID );

    canvas.style.width  = w + "px";
    canvas.style.height = h + "px";
    canvas.width = w;
    canvas.height = h;

    var context = canvas.getContext("2d");
    context.clearRect( 0, 0, w, h );
    context.drawImage( img, 0, 0 );

	if ( isNaN(radius) || radius < 1 ) return;

	if ( blurAlphaChannel )
	{
		boxBlurCanvasRGBA( canvasID, 0, 0, w, h, radius, iterations );
	} else {
		boxBlurCanvasRGB( canvasID, 0, 0, w, h, radius, iterations );
	}

}


function boxBlurCanvasRGBA( id, top_x, top_y, width, height, radius, iterations ){
	if ( isNaN(radius) || radius < 1 ) return;

	radius |= 0;

	if ( isNaN(iterations) ) iterations = 1;
	iterations |= 0;
	if ( iterations > 3 ) iterations = 3;
	if ( iterations < 1 ) iterations = 1;

	var canvas  = document.getElementById( id );
	var context = canvas.getContext("2d");
	var imageData;

	try {
	  try {
		imageData = context.getImageData( top_x, top_y, width, height );
	  } catch(e) {

		// NOTE: this part is supposedly only needed if you want to work with local files
		// so it might be okay to remove the whole try/catch block and just use
		// imageData = context.getImageData( top_x, top_y, width, height );
		try {
			netscape.security.PrivilegeManager.enablePrivilege("UniversalBrowserRead");
			imageData = context.getImageData( top_x, top_y, width, height );
		} catch(e) {
			alert("Cannot access local image");
			throw new Error("unable to access local image data: " + e);
			return;
		}
	  }
	} catch(e) {
	  alert("Cannot access image");
	  throw new Error("unable to access image data: " + e);
	  return;
	}

	var pixels = imageData.data;

	var rsum,gsum,bsum,asum,x,y,i,p,p1,p2,yp,yi,yw,idx,pa;
	var wm = width - 1;
  	var hm = height - 1;
    var wh = width * height;
	var rad1 = radius + 1;

	var mul_sum = mul_table[radius];
	var shg_sum = shg_table[radius];

	var r = [];
    var g = [];
    var b = [];
	var a = [];

	var vmin = [];
	var vmax = [];

	while ( iterations-- > 0 ){
		yw = yi = 0;

		for ( y=0; y < height; y++ ){
			rsum = pixels[yw]   * rad1;
			gsum = pixels[yw+1] * rad1;
			bsum = pixels[yw+2] * rad1;
			asum = pixels[yw+3] * rad1;


			for( i = 1; i <= radius; i++ ){
				p = yw + (((i > wm ? wm : i )) << 2 );
				rsum += pixels[p++];
				gsum += pixels[p++];
				bsum += pixels[p++];
				asum += pixels[p]
			}

			for ( x = 0; x < width; x++ ) {
				r[yi] = rsum;
				g[yi] = gsum;
				b[yi] = bsum;
				a[yi] = asum;

				if( y==0) {
					vmin[x] = ( ( p = x + rad1) < wm ? p : wm ) << 2;
					vmax[x] = ( ( p = x - radius) > 0 ? p << 2 : 0 );
				}

				p1 = yw + vmin[x];
				p2 = yw + vmax[x];

				rsum += pixels[p1++] - pixels[p2++];
				gsum += pixels[p1++] - pixels[p2++];
				bsum += pixels[p1++] - pixels[p2++];
				asum += pixels[p1]   - pixels[p2];

				yi++;
			}
			yw += ( width << 2 );
		}

		for ( x = 0; x < width; x++ ) {
			yp = x;
			rsum = r[yp] * rad1;
			gsum = g[yp] * rad1;
			bsum = b[yp] * rad1;
			asum = a[yp] * rad1;

			for( i = 1; i <= radius; i++ ) {
			  yp += ( i > hm ? 0 : width );
			  rsum += r[yp];
			  gsum += g[yp];
			  bsum += b[yp];
			  asum += a[yp];
			}

			yi = x << 2;
			for ( y = 0; y < height; y++) {

				pixels[yi+3] = pa = (asum * mul_sum) >>> shg_sum;
				if ( pa > 0 )
				{
					pa = 255 / pa;
					pixels[yi]   = ((rsum * mul_sum) >>> shg_sum) * pa;
					pixels[yi+1] = ((gsum * mul_sum) >>> shg_sum) * pa;
					pixels[yi+2] = ((bsum * mul_sum) >>> shg_sum) * pa;
				} else {
					pixels[yi] = pixels[yi+1] = pixels[yi+2] = 0;
				}
				if( x == 0 ) {
					vmin[y] = ( ( p = y + rad1) < hm ? p : hm ) * width;
					vmax[y] = ( ( p = y - radius) > 0 ? p * width : 0 );
				}

				p1 = x + vmin[y];
				p2 = x + vmax[y];

				rsum += r[p1] - r[p2];
				gsum += g[p1] - g[p2];
				bsum += b[p1] - b[p2];
				asum += a[p1] - a[p2];

				yi += width << 2;
			}
		}
	}

	context.putImageData( imageData, top_x, top_y );

}

function boxBlurCanvasRGB( id, top_x, top_y, width, height, radius, iterations ){
	if ( isNaN(radius) || radius < 1 ) return;

	radius |= 0;

	if ( isNaN(iterations) ) iterations = 1;
	iterations |= 0;
	if ( iterations > 3 ) iterations = 3;
	if ( iterations < 1 ) iterations = 1;

	var canvas  = id;//document.getElementById( id );
	var context = canvas.getContext("2d");
	var imageData;

	try {
	  try {
		imageData = context.getImageData( top_x, top_y, width, height );
	  } catch(e) {

		// NOTE: this part is supposedly only needed if you want to work with local files
		// so it might be okay to remove the whole try/catch block and just use
		// imageData = context.getImageData( top_x, top_y, width, height );
		try {
			netscape.security.PrivilegeManager.enablePrivilege("UniversalBrowserRead");
			imageData = context.getImageData( top_x, top_y, width, height );
		} catch(e) {
			alert("Cannot access local image");
			throw new Error("unable to access local image data: " + e);
			return;
		}
	  }
	} catch(e) {
	  alert("Cannot access image");
	  throw new Error("unable to access image data: " + e);
	  return;
	}

	var pixels = imageData.data;

	var rsum,gsum,bsum,asum,x,y,i,p,p1,p2,yp,yi,yw,idx;
	var wm = width - 1;
  	var hm = height - 1;
    var wh = width * height;
	var rad1 = radius + 1;

	var r = [];
    var g = [];
    var b = [];

	var mul_sum = mul_table[radius];
	var shg_sum = shg_table[radius];

	var vmin = [];
	var vmax = [];

	while ( iterations-- > 0 ){
		yw = yi = 0;

		for ( y=0; y < height; y++ ){
			rsum = pixels[yw]   * rad1;
			gsum = pixels[yw+1] * rad1;
			bsum = pixels[yw+2] * rad1;

			for( i = 1; i <= radius; i++ ){
				p = yw + (((i > wm ? wm : i )) << 2 );
				rsum += pixels[p++];
				gsum += pixels[p++];
				bsum += pixels[p++];
			}

			for ( x = 0; x < width; x++ ){
				r[yi] = rsum;
				g[yi] = gsum;
				b[yi] = bsum;

				if( y==0) {
					vmin[x] = ( ( p = x + rad1) < wm ? p : wm ) << 2;
					vmax[x] = ( ( p = x - radius) > 0 ? p << 2 : 0 );
				}

				p1 = yw + vmin[x];
				p2 = yw + vmax[x];

				rsum += pixels[p1++] - pixels[p2++];
				gsum += pixels[p1++] - pixels[p2++];
				bsum += pixels[p1++] - pixels[p2++];

				yi++;
			}
			yw += ( width << 2 );
		}

		for ( x = 0; x < width; x++ ){
			yp = x;
			rsum = r[yp] * rad1;
			gsum = g[yp] * rad1;
			bsum = b[yp] * rad1;

			for( i = 1; i <= radius; i++ ){
			  yp += ( i > hm ? 0 : width );
			  rsum += r[yp];
			  gsum += g[yp];
			  bsum += b[yp];
			}

			yi = x << 2;
			for ( y = 0; y < height; y++){
				pixels[yi]   = (rsum * mul_sum) >>> shg_sum;
				pixels[yi+1] = (gsum * mul_sum) >>> shg_sum;
				pixels[yi+2] = (bsum * mul_sum) >>> shg_sum;

				if( x == 0 ) {
					vmin[y] = ( ( p = y + rad1) < hm ? p : hm ) * width;
					vmax[y] = ( ( p = y - radius) > 0 ? p * width : 0 );
				}

				p1 = x + vmin[y];
				p2 = x + vmax[y];

				rsum += r[p1] - r[p2];
				gsum += g[p1] - g[p2];
				bsum += b[p1] - b[p2];

				yi += width << 2;
			}
		}
	}
	context.putImageData( imageData, top_x, top_y );

}